_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+
+# quadratic penalty
+penalty <- function(x) torch_square(torch_max(constraint(x), other = 0))
+
+```
+
+A priori, we can't know how big that multiple has to be to enforce the constraint. Therefore, optimization proceeds iteratively. We start with a small multiplier, $1$, say, and increase it for as long as the constraint is still violated:
+
+```{r}
+penalty_method <- function(f, p, x, k_max, rho = 1, gamma = 2, num_iterations = 1) {
+
+  for (k in 1:k_max) {
+    cat("Starting step: ", k, ", rho = ", rho, "\n")
+
+    minimize(f, p, x, rho, num_iterations)
+
+    cat("Value: ",  as.numeric(f(x)), "\n")
+    cat("X: ",  as.matrix(x), "\n")
+    
+    current_penalty <- as.numeric(p(x))
+    cat("Penalty: ", current_penalty, "\n")
+    if (current_penalty == 0) break
+    
+    rho <- rho * gamma
+  }
+
+}
+```
+
+`Minimize()`, called from `penalty_method()`, follows the usual proceedings, but now it minimizes the sum of the target and up-weighted penalty function outputs:
+
+```{r}
+minimize <- function(f, p, x, rho, num_iterations) {
+
+  calc_loss <- function() {
+    optimizer$zero_grad()
+    value <- f(x) + rho * p(x)
+    value$backward()
+    value
+  }
+
+  for (i in 1:num_iterations) {
+    cat("Iteration: ", i, "\n")
+    optimizer$step(calc_loss)
+  }
+
+}
+```
+
+This time, we start from a low-target-loss, but unfeasible value. With yet another change to default L-BFGS (namely, a decrease in tolerance), we see the algorithm exiting successfully after twenty-two iterations, at the point `(0.5411692,1.306563)`.
+
+```{r}
+x_star <- torch_tensor(c(0.5, 0.5), requires_grad = TRUE)
+
+optimizer <- optim_lbfgs(x_star, line_search_fn = "strong_wolfe", tolerance_change = 1e-20)
+
+penalty_method(flower, penalty, x_star, k_max = 30)
+
+```
+
+    Starting step:  1 , rho =  1 
+    Iteration:  1 
+    Value:  0.3469974 
+    X:  0.5154735 1.244463 
+    Penalty:  0.03444662 
+
+    Starting step:  2 , rho =  2 
+    Iteration:  1 
+    Value:  0.3818618 
+    X:  0.5288152 1.276674 
+    Penalty:  0.008182613 
+
+    Starting step:  3 , rho =  4 
+    Iteration:  1 
+    Value:  0.3983252 
+    X:  0.5351116 1.291886 
+    Penalty:  0.001996888 
+
+    ...
+    ...
+
+    Starting step:  20 , rho =  524288 
+    Iteration:  1 
+    Value:  0.4142133 
+    X:  0.5411959 1.306563 
+    Penalty:  3.552714e-13 
+
+    Starting step:  21 , rho =  1048576 
+    Iteration:  1 
+    Value:  0.4142134 
+    X:  0.5411956 1.306563 
+    Penalty:  1.278977e-13 
+
+    Starting step:  22 , rho =  2097152 
+    Iteration:  1 
+    Value:  0.4142135 
+    X:  0.5411962 1.306563 
+    Penalty:  0 
+
+## Conclusion
+
+Summing up, we've gotten a first impression of the effectiveness of `torch`'s L-BFGS optimizer, especially when used with Strong-Wolfe line search. In fact, in numerical optimization -- as opposed to deep learning, where computational speed is much more of an issue -- there probably is hardly ever a reason to *not* use L-BFGS with line search.
```

Comment:
```suggestion
Summing up, we've gotten a first impression of the effectiveness of `torch`'s L-BFGS optimizer, especially when used with Strong-Wolfe line search. In fact, in numerical optimization -- as opposed to deep learning, where computational speed is much more of an issue -- there is hardly ever a reason to *not* use L-BFGS with line search.
```

Reply:
You don't need probably when you've got "hardly ever," since it's akin to equivocating/qualifying twice.

Reply:
caught me again (you know what I'm referring to :-))